File: leetcode/problems/cpp/longest-substring-without-repeating-characters.cpp
// by @codeAbinash
// Time : O(n)
// Space : O(1)

#include "string"
#include "vector"
using namespace std;

class Solution {
public:
   int lengthOfLongestSubstring(string s) {
      bool set[128] = { false };
      int maxLen = 0, len = s.length(), left = 0;

      for (int i = 0; i < len; i++) {
         while (set[s[i]])
            set[s[left]] = false, left++;

         set[s[i]] = true;
         maxLen = max(maxLen, i - left + 1);
      }
      return maxLen;
   }
};

File: leetcode/problems/rs/longest-substring-without-repeating-characters.rs
// by @codeAbinash
// Time : O(n)
// Space : O(n)

pub struct Solution;
use std::collections::VecDeque;
use std::cmp::max;

impl Solution {
   pub fn length_of_longest_substring(s: String) -> i32 {
      let mut result = 0;
      let mut window = VecDeque::<char>::new();

      for ch in s.chars() {
         if window.contains(&ch) {
            while let Some(prev) = window.pop_front() {
               if prev == ch {
                  break;
               }
            }
         }

         window.push_back(ch);
         result = max(result, window.len());
      }

      result as i32
   }
}
